README.md
<h2 align="center">Great Efue Date Wizard</h2>

<p align="center">A Package which transform date to a friendly sentence etc.</p>

<p align="center">
    <a href="https://www.npmjs.com/package/greatefue-datewizard"><img src="https://img.shields.io/npm/v/greatefue-datewizard.svg?style=flat-square" alt="NPM Version" /></a>
    <a href="https://www.npmjs.com/package/greatefue-datewizard"><img src="https://img.shields.io/npm/dm/greatefue-datewizard.svg?style=flat-square" alt="NPM Downloads" /></a>
    <img src="https://github.com/greatefue/greatefue-datewizard/workflows/CI/badge.svg" alt="CI" />
    <a href="https://github.com/greatefue/greatefue-datewizard/tree/1.0.4"></a>
    <a href="https://app.fossa.com/api/projects/git%2Bgithub.com%2Fgreatefue%2Fgreatefue-datewizard.svg?type=shield"><img src="https://app.fossa.com/api/projects/git%2Bgithub.com%2Fgreatefue%2Fgreatefue-datewizard.svg?type=shield"/></a>
</p>


## Table of Contents

1. [Installation](#installation)
2. [Usage](#usage)
3. [Optional Parameter Table](#optional-parameter-table)
4. [Releases](./HISTORY.md)
4. [Author](#author)
6. [License](#license)
7. [Team](#team)
<!-- 13. [Technology Sponsors](#technology-sponsors) -->

## <a name="installation"></a>Installation

```bash
$ yarn add greatefue-datewizard
$ npm i greatefue-datewizard
```

## <a name="usage"></a>Usage

In your .ts or .js file, 


1. import the package

```ts
import { DaysAgo } from 'greatefue-datewizard';
```

2. Call the function

```ts
var i = DaysAgo("2021-01-01");
console.log(i) //2 months ago
```

3. Passing Optional Parameters
```ts
var i = DaysAgo("2021-01-01",{date:true});
console.log(i) //2 months ago
```

4. Passing a future date returns difference from the current date to the future date.
```ts
var i = DaysAgo("2099-12-31");
console.log(i) //78 years into the future
```

## <a name="optional-parameter-table"></a>Optional Parameter Table

|**Parameter** | **Datatype** | **Description**                                          | **Example**                                               | **Example Output**  |
| ------------ |:------------:| :------------------------------------------------------- | :-------------------------------------------------------- | :------------------ |
| date         | boolean      | Extract **year month day** from date or datetime         | ``` DaysAgo("2021-01-01T00:00:00",{date:true}); ```       | //2 months ago      |
| time         | boolean      | Extract **hour minute second** from date or datetime     | ``` DaysAgo("2021-01-01T00:00:00",{time:true}); ```       | //2 hours ago       |
| year         | boolean      | Extract **year** from date or datetime                   | ``` DaysAgo("2021-01-01T00:00:00",{year:true}); ```       | //2 years ago       |
| month        | boolean      | Extract **month** from date or datetime                  | ``` DaysAgo("2021-01-01T00:00:00",{month:true}); ```      | //2 months ago      |
| day          | boolean      | Extract **day** from date or datetime                    | ``` DaysAgo("2021-01-01T00:00:00",{day:true}); ```        | //2 days ago        |
| hour         | boolean      | Extract **hour** from datetime                           | ``` DaysAgo("2021-01-01T00:00:00",{hour:true}); ```       | //2 hours ago       |
| minute       | boolean      | Extract **minute** from datetime                         | ``` DaysAgo("2021-01-01T00:00:00",{minute:true}); ```     | //2 minutes ago     |
| second       | boolean      | Extract **second** from datetime                         | ``` DaysAgo("2021-01-01T00:00:30",{second:true}); ```     | //2 second ago      |



## <a name="author"></a>Author

**Great Efue**

* [GitHub Profile](https://github.com/greatefue)
* [Twitter Profile](https://twitter.com/achillesefue)
* [LinkedIn Profile](https://linkedin.com/in/greatefue)


Please consider supporting me on Patreon.
<a href="https://www.patreon.com/bePatron?u=53060354" data-patreon-widget-type="become-patron-button">Become a Patron of Great Efue!</a>


## <a name="license"></a>License

[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fgreatefue%2Fgreatefue-datewizard.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fgreatefue%2Fgreatefue-datewizard?ref=badge_large)


## <a name="team"></a>Team

These are folks that keep the project moving and are resources for help.

### Technical Steering Committee (TSC)

The people who manage releases, review feature requests, and meet regularly to ensure GreatEfue-DateWizard is properly maintained.

<table><tbody><tr><td align="center" valign="top" width="11%">
<a href="https://github.com/greatefue">
<img src="https://github.com/greatefue.png?s=75" width="75" height="75"><br />
Great Efue
</a>
</td><td align="center" valign="top" width="11%">
<a href="https://github.com/d-sonofman">
<img src="https://github.com/d-sonofman.png?s=75" width="75" height="75"><br />
Temitope Afolabi
</a>
</td></tr></tbody></table>


### Reviewers

The people who review and implement new features.

<table><tbody><tr><td align="center" valign="top" width="11%">
<a href="https://github.com/greatefue">
<img src="https://github.com/greatefue.png?s=75" width="75" height="75"><br />
Great Efue
</a>
</td></tr></tbody></table>

### Committers

The people who review and fix bugs and help triage issues.

<table><tbody><tr><td align="center" valign="top" width="11%">
<a href="https://github.com/greatefue">
<img src="https://github.com/greatefue.png?s=75" width="75" height="75"><br />
Great Efue
</a>
</td></tr></tbody></table>



<!-- ## <a name="technology-sponsors"></a>Technology Sponsors

The following companies, organizations, and individuals support GreatEfue-DateWizard ongoing maintenance and development. [Become a Sponsor](https://opencollective.com/greatefue-datewizard) to get your logo on our README and website. -->

<!-- NOTE: This section is autogenerated. Do not manually edit.-->
<!--sponsorsstart-->
<!-- <h3>Platinum Sponsors</h3>
<p><a href="https://solutions.efuelite.com"><img src="https://solutions.efuelite.com/assets/img/EfueliteSolutions_Logo.png" alt="Efuelite Solutions" width="75" height="75"></a></p> -->

<!--sponsorsend-->


HISTORY.md
2021-04-01 / 1.0.4
==================

  * * Fix the way inheriting functions
  * * Modularized functions

2021-03-29 / 1.0.3
==================

  * * Fix date conversion error

2021-03-29 / 1.0.2
==================

  * * Fix date conversion error

2021-03-29 / 1.0.1
==================

  * Fix date conversion error

2021-03-25 / 1.0.0
==================

  * Initial release


index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DaysAgo = void 0;
const variables_1 = require("./variables");
function DaysAgo(value, DateInstructions = {}) {
    if (value) {
        if (!Date.parse(value.toString())) {
            return value + " " + "is an invalid date.";
        }
        const IsInstructionEmpty = ObjectAnalyzer.IsObjectEmpty(DateInstructions);
        let newDate = +new Date();
        const intervals = variables_1.PreDefinedVariables.DateIntervals;
        const InstructionUsed = DatePart.ExtractDatePart(value, DateInstructions).InstructionUsed;
        if (!IsInstructionEmpty) {
            value = DatePart.ExtractDatePart(value, DateInstructions).value;
            newDate = DatePart.ExtractDatePart(+new Date(), DateInstructions).value;
        }
        const seconds = Math.floor((+new Date(newDate) - +new Date(value)) / 1000);
        if (seconds > -1 && seconds < 29 && InstructionUsed == "none") {
            return "Just now";
        }
        if (seconds === 0 && InstructionUsed == "date") {
            return "This day";
        }
        let counter;
        if (seconds < 0 && InstructionUsed === "date" || InstructionUsed === "time" || InstructionUsed === "none") {
            const seconds = Math.floor((+new Date(value) - +new Date(newDate)) / 1000);
            for (const i in intervals) {
                counter = Math.floor(seconds / intervals[i]);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + i + " into the future";
                    }
                    else {
                        return counter + " " + i + "s into the future";
                    }
                }
            }
        }
        if (seconds > 0 && InstructionUsed === "date" || InstructionUsed === "time" || InstructionUsed === "none") {
            for (const i in intervals) {
                counter = Math.floor(seconds / intervals[i]);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + i + " ago";
                    }
                    else {
                        return counter + " " + i + "s ago";
                    }
                }
            }
        }
        if (InstructionUsed === "year") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "month") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "day") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "hour") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "minute") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "second") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
            else {
                counter = +new Date(value) - +new Date(newDate);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                }
                else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
    }
    return value;
}
exports.DaysAgo = DaysAgo;
class DatePart {
    static ExtractDatePart(value, instructions) {
        if ((Object.getOwnPropertyDescriptor(instructions, "date")) && instructions.date === true) {
            const v = (new Date(value).getFullYear()) + "-" + (new Date(value).getMonth() + 1) + "-" + (new Date(value).getDate());
            return { value: +new Date(v), InstructionUsed: "date" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "time")) && instructions.time === true) {
            const v = +new Date(value).getHours() + "-" + +new Date(value).getMinutes() + "-" + +new Date(value).getSeconds();
            return { value: +new Date(v), InstructionUsed: "time" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "year")) && instructions.year === true) {
            return { value: +new Date(value).getFullYear(), InstructionUsed: "year" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "month")) && instructions.month === true) {
            return { value: +new Date(value).getMonth() + 1, InstructionUsed: "month" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "day")) && instructions.day === true) {
            return { value: +new Date(value).getDate(), InstructionUsed: "day" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "hour")) && instructions.hour === true) {
            return { value: +new Date(value).getHours(), InstructionUsed: "hour" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "minute")) && instructions.minute === true) {
            return { value: +new Date(value).getMinutes(), InstructionUsed: "minute" };
        }
        if ((Object.getOwnPropertyDescriptor(instructions, "second")) && instructions.second === true) {
            return { value: +new Date(value).getSeconds(), InstructionUsed: "second" };
        }
        return { value: +new Date(value), InstructionUsed: "none" };
    }
}
class ObjectAnalyzer {
    static IsObjectEmpty(obj) {
        if (Object.getOwnPropertyNames(obj).length === 0) {
            return true;
        }
        return false;
    }
}


index.ts

type DateInstructionsOptions = {
    [key: string]: boolean
}
export const DateInstructions = {
    "year": false,
    "month": false,
    "day": false,
    "hour": false,
    "minute": false,
    "second": false,
    "time": false,
    "date": false
};
export function DaysAgo(value: string | number | Date, DateInstructions: DateInstructionsOptions = {}): string | number | Date {

    if (value) {

        if (!Date.parse(value.toString())) {
            return value + " " + "is an invalid date.";
        }

        const IsInstructionEmpty: boolean = IsObjectEmpty(DateInstructions);

        let newDate = +new Date();
        type intervalsOptions = {
            [key: string]: number
        }

        const intervals: intervalsOptions = {
            "year": 31536000,
            "month": 2592000,
            "week": 604800,
            "day": 86400,
            "hour": 3600,
            "minute": 60,
            "second": 1
        };

        const InstructionUsed = ExtractDatePart(value, DateInstructions).InstructionUsed;

        if (!IsInstructionEmpty) {
            value = ExtractDatePart(value, DateInstructions).value;
            newDate = ExtractDatePart(+new Date(), DateInstructions).value;
        }
        const seconds = Math.floor((+new Date(newDate) - +new Date(value)) / 1000);
        if (seconds > -1 && seconds < 29 && InstructionUsed == "none") {
            return "Just now";
        }

        if (seconds === 0 && InstructionUsed == "date") {
            return "This day";
        }

        let counter;
        if (seconds < 0 && InstructionUsed === "date" || InstructionUsed === "time" || InstructionUsed === "none") {
            const seconds = Math.floor((+new Date(value) - +new Date(newDate)) / 1000);
            for (const i in intervals) {
                counter = Math.floor(seconds / intervals[i]);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + i + " into the future";
                    }
                    else {
                        return counter + " " + i + "s into the future";
                    }
                }
            }
        }
        if (seconds > 0 && InstructionUsed === "date" || InstructionUsed === "time" || InstructionUsed === "none") {
            for (const i in intervals) {
                counter = Math.floor(seconds / intervals[i]);
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + i + " ago";
                    }
                    else {
                        return counter + " " + i + "s ago";
                    }
                }
            }
        }
        if (InstructionUsed === "year") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "month") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "day") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "hour") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "minute") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }
        if (InstructionUsed === "second") {
            if ((+new Date(newDate) > +new Date(value))) {
                counter = +new Date(newDate) - +new Date(value)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " ago";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s ago";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            } else {
                counter = +new Date(value) - +new Date(newDate)
                if (counter > 0) {
                    if (counter === 1) {
                        return counter + " " + InstructionUsed + " into the future";
                    }
                    else {
                        return counter + " " + InstructionUsed + "s into the future";
                    }
                } else {
                    return "This" + " " + InstructionUsed;
                }
            }
        }

    }
    return value;
}


function ExtractDatePart(value: string | number | Date, instructions: DateInstructionsOptions) {
    if ((Object.getOwnPropertyDescriptor(instructions, "date")) && instructions.date === true) {
        const v = (new Date(value).getFullYear()) + "-" + (new Date(value).getMonth() + 1) + "-" + (new Date(value).getDate());
        return { value: +new Date(v), InstructionUsed: "date" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "time")) && instructions.time === true) {
        const v = +new Date(value).getHours() + "-" + +new Date(value).getMinutes() + "-" + +new Date(value).getSeconds();
        return { value: +new Date(v), InstructionUsed: "time" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "year")) && instructions.year === true) {
        return { value: +new Date(value).getFullYear(), InstructionUsed: "year" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "month")) && instructions.month === true) {
        return { value: +new Date(value).getMonth() + 1, InstructionUsed: "month" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "day")) && instructions.day === true) {
        return { value: +new Date(value).getDate(), InstructionUsed: "day" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "hour")) && instructions.hour === true) {
        return { value: +new Date(value).getHours(), InstructionUsed: "hour" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "minute")) && instructions.minute === true) {
        return { value: +new Date(value).getMinutes(), InstructionUsed: "minute" };
    }
    if ((Object.getOwnPropertyDescriptor(instructions, "second")) && instructions.second === true) {
        return { value: +new Date(value).getSeconds(), InstructionUsed: "second" };
    }
    return { value: +new Date(value), InstructionUsed: "none" };
}

function IsObjectEmpty(obj: DateInstructionsOptions) {
    if (Object.getOwnPropertyNames(obj).length === 0) {
        return true;
    }
    return false;
}

